src/ucs/sys/ucs_netlink.c

```diff
@@ -0,0 +1,255 @@
+    nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+    iov[0].iov_base = &nlh;
+    iov[0].iov_len  = sizeof(nlh);
+    iov[1].iov_base = nl_protocol_hdr;
+    iov[1].iov_len  = nl_protocol_hdr_size;
+
+    do {
+        status = ucs_socket_sendv_nb(fd, iov, 2, &dummy);
+    } while (status == UCS_ERR_NO_PROGRESS);
+
+    if (status != UCS_OK) {
+        ucs_error("failed to send netlink message (%s)",
+                  ucs_status_string(status));
+        goto out;
+    }
+
+    do {
+        status = ucs_socket_recv_nb(fd, recv_msg_buf, recv_msg_buf_len);
+    } while (status == UCS_ERR_NO_PROGRESS);
+
+    if (status != UCS_OK) {
+        ucs_error("failed to receive netlink message");
+        goto out;
+    }
+
+out:
+    ucs_close_fd(&fd);
+    return status;
+}
+
+ucs_status_t ucs_netlink_parse_msg(void *msg, size_t msg_len,
+                                   ucs_netlink_parse_cb_t parse_cb, void *arg)
+{
+    struct nlmsghdr *nlh;
+    ucs_status_t status = UCS_INPROGRESS;
+
+    for (nlh = (struct nlmsghdr*)msg;
+         (status == UCS_INPROGRESS) && NLMSG_OK(nlh, msg_len) &&
+         (nlh->nlmsg_type != NLMSG_DONE) && (nlh->nlmsg_type != NLMSG_ERROR);
+         nlh = NLMSG_NEXT(nlh, msg_len)) {
+        status = parse_cb(nlh, NLMSG_DATA(nlh), arg);
+    }
+
+    if (nlh->nlmsg_type == NLMSG_ERROR) {
+        struct nlmsgerr *err = (struct nlmsgerr*)NLMSG_DATA(nlh);
+        ucs_error("failed to parse netlink message header (%d)", err->error);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+
+/* *******************************************************
+ * Route Netlink utilities                               *
+ * ***************************************************** */
+
+#define NETLINK_MESSAGE_MAX_SIZE 8195
+
+struct route_info {
+    const struct sockaddr *sa_remote;
+    int                    if_index;
+    int                    matching;
+};
+
+
+static ucs_status_t ucs_rtnetlink_get_route_info(int *if_idx, void **dst_in_addr,
+                                                 struct rtattr *rta, int len)
+{
+    *if_idx      = -1;
+    *dst_in_addr = NULL;
+
+    for (; RTA_OK(rta, len); rta = RTA_NEXT(rta, len)) {
+        if (rta->rta_type == RTA_OIF) {
+            *if_idx = *((int*)RTA_DATA(rta));
+        } else if (rta->rta_type == RTA_DST) {
+            *dst_in_addr = RTA_DATA(rta);
+        }
+    }
+
+    if ((*if_idx == -1) || (*dst_in_addr == NULL)) {
+        ucs_debug("either iface index or dest address are missing "
+                  "in the routing table entry");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+static int ucs_rtnetlink_is_rule_matching(struct rtmsg *rtm, size_t rtm_len,
+                                          const struct sockaddr *sa_remote,
+                                          int iface_index)
+{
+    int   rule_iface;
+    void *dst_in_addr;
+
+    if (rtm->rtm_family != sa_remote->sa_family) {
+        return 0;
+    }
```

Comment:
address family is guaranteed to match because that's what we passed in the netlink request (line 214)